devito/data/utils.py

```diff
@@ -0,0 +1,330 @@
+        left_neighbours = as_tuple(left_neighbours)
+        n_dat = []  # Normalised data size
+        for j in range(len(my_coords)):
+            if left_neighbours[j][j] < 0:
+                c_dat = dat_len[my_coords][j]
+                n_dat.append(c_dat)
+            else:
+                c_dat = dat_len[my_coords][j]  # Current length
+                p_dat = dat_len[left_neighbours[j]][j]  # Previous length
+                n_dat.append(c_dat+p_dat)
+        dat_len_cum[my_coords] = as_tuple(n_dat)
+    # This 'transform' will be required to produce the required maps
+    transform = []
+    for i in as_tuple(loc_idx):
+        if isinstance(i, slice):
+            if i.step is not None:
+                transform.append(slice(None, None, np.sign(i.step)))
+            else:
+                transform.append(slice(None, None, None))
+        else:
+            transform.append(0)
+    transform = as_tuple(transform)
+
+    global_size = dat_len_cum[coords[-1]]
+
+    indices = np.zeros(global_size, dtype=tuple)
+    global_si = np.zeros(global_size, dtype=tuple)
+    it = np.nditer(indices, flags=['refs_ok', 'multi_index'])
+    while not it.finished:
+        index = it.multi_index
+        indices[index] = index
+        it.iternext()
+    global_si[:] = indices[transform]
+
+    # Create the 'rank' slices
+    rank_slice = []
+    for j in coords:
+        this_rank = []
+        for k in dat_len[j]:
+            this_rank.append(slice(0, k, 1))
+        rank_slice.append(this_rank)
+    # Normalize the slices:
+    n_rank_slice = []
+    for i in range(len(rank_slice)):
+        my_coords = coords[i]
+        if any([j.stop == j.start for j in rank_slice[i]]):
+            n_rank_slice.append(as_tuple([None]*len(rank_slice[i])))
+            continue
+        if i == 0:
+            n_rank_slice.append(as_tuple(rank_slice[i]))
+            continue
+        left_neighbours = []
+        for j in range(len(my_coords)):
+            left_coord = list(my_coords)
+            left_coord[j] -= 1
+            left_neighbours.append(as_tuple(left_coord))
+        left_neighbours = as_tuple(left_neighbours)
+        n_slice = []
+        for j in range(len(my_coords)):
+            if left_neighbours[j][j] < 0:
+                start = 0
+                stop = dat_len_cum[my_coords][j]
+            else:
+                start = dat_len_cum[left_neighbours[j]][j]
+                stop = dat_len_cum[my_coords][j]
+            n_slice.append(slice(start, stop, 1))
+        n_rank_slice.append(as_tuple(n_slice))
+    n_rank_slice = as_tuple(n_rank_slice)
+
+    # Now fill each elements owner:
+    owners = np.zeros(global_size, dtype=np.int32)
+    send = np.zeros(global_size, dtype=np.int32)
+    for i in range(len(n_rank_slice)):
+        if any([j is None for j in n_rank_slice[i]]):
+            continue
+        else:
+            owners[n_rank_slice[i]] = i
+    send[:] = owners[transform]
+
+    # local_indices
```

Comment:
Can you elaborate -- local_indices what? you mean "Create local indices" ?

Reply:
It's constructing `local_si` which has been described in the docsting. Changed comment to be more appropriate.